## Load Data

In [4]:
from src.morse import Morse
from src.evaluation_and_conversion import compare_result_txt_to_groundtruth_ply
import itertools

In [8]:
def process(artifact, persistence, params, index = 7):
    data_conf = Morse()
    data_n = Morse()
    data_conf.load_mesh_ply("Data/"+artifact+".ply", quality_index=index, inverted=True)
    data_n.load_mesh_ply("Data/"+artifact+".ply", quality_index=index, inverted=True)

    outfilename =  "Data/results/" + artifact + "/" + artifact

    #### CONFORMING
    data_conf.load_labels("Data/"+artifact+"_labels.txt")
    data_conf.ConformingGradient()
    data_conf.ExtractMorseComplex()
    data_conf.ReduceMorseComplex(10, conforming=True)

    accuracy = 0
    best_params = []

    for pers in persistence:
        data_conf.ReduceMorseComplex(pers, conforming=True)
        data_conf.ExtractMorseCells(pers)

        for high, low, merge in params:
            if high > low:
                data_conf.Segmentation(pers, high, low, merge, conforming=True)
                data_conf.plot_Segmentation_label_txt(pers, high, low, merge, outfilename)
                result = compare_result_txt_to_groundtruth_ply(outfilename+ "_" + str(pers) + "P_" + str(high) + "-" + str(low) 
                                       + "T_" + str(merge) + ".txt", "Data/"+artifact+"_c.ply")
                if result[0] > accuracy:
                    accuracy = result[0]
                    best_params = result[1:]

    print(accuracy, best_params)

    ### NON CONFORMING
    data_n.ProcessLowerStars()
    data_n.ExtractMorseComplex()

    data_n.ReduceMorseComplex(best_params[0], conforming=False)
    data_n.ExtractMorseCells(best_params[0])
    data_n.Segmentation(best_params[0], best_params[1], best_params[2], best_params[3], conforming=False)
    data_n.plot_Segmentation_label_txt(best_params[0], best_params[1], best_params[2], best_params[3], outfilename+"non_conf")




In [9]:
# PARAMETERS
persistence = [0.03, 0.04]
high_thresh = [0.05, 0.06, 0.07]
low_thresh = [0.01, 0.02]
merge_thresh = [0.3, 0.4, 0.5]


process("700", persistence, list(itertools.product(high_thresh, low_thresh, merge_thresh)))

load_mesh_ply took 18.42344 seconds to execute!
load_mesh_ply took 21.16393 seconds to execute!
Time read labels in txt file: 0.1387857999652624
Time ProcessLowerStar: 6.494311399990693
ExtractMorseComplex took 4.59522 seconds to execute!
Persistence was high enough that this complex is maximally reduced.
ReduceMorseComplex took 5.19965 seconds to execute!
ReduceMorseComplex took 9.39584 seconds to execute!
BD points:  41670
ExtractMorseCells took 2.42277 seconds to execute!
get_salient_edges took 0.10771 seconds to execute!
Segmentation took 18.32863 seconds to execute!
plot_Segmentation_label_txt took 0.32015 seconds to execute!
compare_result_txt_to_groundtruth_ply took 16.72953 seconds to execute!
get_salient_edges took 0.15858 seconds to execute!
Segmentation took 22.03692 seconds to execute!
plot_Segmentation_label_txt took 0.23736 seconds to execute!
compare_result_txt_to_groundtruth_ply took 11.64690 seconds to execute!
get_salient_edges took 0.15359 seconds to execute!
Segment

In [4]:
# PARAMETERS
persistence = [0.03, 0.04, 0.05]
high_thresh = [0.04, 0.05, 0.06]
low_thresh = [0.02, 0.03, 0.04]
merge_thresh = [0.1, 0.2, 0.3]

process("801", persistence, list(itertools.product(high_thresh, low_thresh, merge_thresh)), index = 3)

load_mesh_ply took 21.44164 seconds to execute!
load_mesh_ply took 20.88740 seconds to execute!
Time read labels in txt file: 0.11172439996153116
Time ProcessLowerStar: 6.2518251000437886
ExtractMorseComplex took 6.34803 seconds to execute!
ReduceMorseComplex took 4.09324 seconds to execute!
BD points:  53106
ExtractMorseCells took 2.00489 seconds to execute!
Need maximally reduced complex for salient edges...
Persistence was high enough that this complex is maximally reduced.
ReduceMorseComplex took 4.40124 seconds to execute!
get_salient_edges took 0.08078 seconds to execute!
Segmentation took 15.56103 seconds to execute!
plot_Segmentation_label_txt took 0.22241 seconds to execute!
compare_result_txt_to_groundtruth_ply took 9.57789 seconds to execute!
get_salient_edges took 0.09275 seconds to execute!
Segmentation took 14.23113 seconds to execute!
plot_Segmentation_label_txt took 0.23138 seconds to execute!
compare_result_txt_to_groundtruth_ply took 9.55283 seconds to execute!
get_sa

In [5]:
# PARAMETERS
persistence = [0.06, 0.07, 0.08]
high_thresh = [0.08, 0.09, 0.1]
low_thresh = [0.06, 0.07, 0.08]
merge_thresh = [0.2, 0.3, 0.4]

process("4501", persistence, list(itertools.product(high_thresh, low_thresh, merge_thresh)))

load_mesh_ply took 17.26305 seconds to execute!
load_mesh_ply took 19.93653 seconds to execute!
Time read labels in txt file: 0.12839339999482036
Time ProcessLowerStar: 7.467729300027713
ExtractMorseComplex took 4.94065 seconds to execute!
ReduceMorseComplex took 9.25874 seconds to execute!
BD points:  43069
ExtractMorseCells took 4.53274 seconds to execute!
Need maximally reduced complex for salient edges...
Persistence was high enough that this complex is maximally reduced.
ReduceMorseComplex took 6.37387 seconds to execute!
get_salient_edges took 0.04687 seconds to execute!
Segmentation took 25.34427 seconds to execute!
plot_Segmentation_label_txt took 0.24138 seconds to execute!
compare_result_txt_to_groundtruth_ply took 10.88471 seconds to execute!
get_salient_edges took 0.04987 seconds to execute!
Segmentation took 20.42887 seconds to execute!
plot_Segmentation_label_txt took 0.24498 seconds to execute!
compare_result_txt_to_groundtruth_ply took 10.42162 seconds to execute!
get_s

In [6]:
print('700:')
result = compare_result_txt_to_groundtruth_ply("Data/results/700/700_0.03P_0.06-0.02T_0.4.txt", "Data/700_c.ply")  
print(result[0])   
result = compare_result_txt_to_groundtruth_ply("Data/results/700/700non_conf_0.03P_0.06-0.02T_0.4.txt", "Data/700_c.ply")  
print(result[0])  

print('801:')
result = compare_result_txt_to_groundtruth_ply("Data/results/801/801_0.04P_0.06-0.04T_0.2.txt", "Data/801_c.ply")     
print(result[0])  
result = compare_result_txt_to_groundtruth_ply("Data/results/801/801non_conf_0.04P_0.06-0.04T_0.2.txt", "Data/801_c.ply")          
print(result[0])  

print('4501:')
result = compare_result_txt_to_groundtruth_ply("Data/results/4501/4501_0.07P_0.1-0.07T_0.3.txt", "Data/4501_c.ply")     
print(result[0])  
result = compare_result_txt_to_groundtruth_ply("Data/results/4501/4501non_conf_0.07P_0.1-0.07T_0.3.txt", "Data/4501_c.ply")    
print(result[0])   

700:
compare_result_txt_to_groundtruth_ply took 9.29391 seconds to execute!
92.49944870998135
compare_result_txt_to_groundtruth_ply took 12.57962 seconds to execute!
73.77363029489004
801:
compare_result_txt_to_groundtruth_ply took 9.04221 seconds to execute!
94.85600282446299
compare_result_txt_to_groundtruth_ply took 9.05347 seconds to execute!
79.80845209340902
4501:
compare_result_txt_to_groundtruth_ply took 10.38498 seconds to execute!
98.02251977480225
compare_result_txt_to_groundtruth_ply took 10.07311 seconds to execute!
97.72552274477255


## Results

### 700
Persistence: 0.03

Thresh_high: 0.06

Thresh_low: 0.02

Thresh_merge: 0.4


Accuracy conforming: 92.5 %

Accuracy non conforming: 73.77 %

### 801
Persistence: 0.04

Thresh_high: 0.06

Thresh_low: 0.04

Thresh_merge: 0.2


Accuracy conforming: 94.86 %

Accuracy non conforming: 79.81 %

### 4501
Persistence: 0.07

Thresh_high: 0.1

Thresh_low: 0.07

Thresh_merge: 0.3


Accuracy conforming: 98.02 %

Accuracy non conforming: 97.73 %

In [22]:
from src.Algorithms.EdgeDetection import get_salient_edge_indices

def write_header(file, nb_points, nb_faces):
    file.write("ply \n")
    file.write("format ascii 1.0 \n")
    file.write("element vertex " + str(nb_points) + "\n")
    file.write("property float x\n")
    file.write("property float y\n")
    file.write("property float z\n")
    file.write("property uchar red\n")
    file.write("property uchar green\n")
    file.write("property uchar blue\n")
    file.write("element face " + str(nb_faces) + "\n")
    file.write("property list uchar int vertex_indices\n")
    file.write("end_header\n")
    
def write_vertex(file, vert, vert_dict, color=[0,0,0]):
    file.write(str(vert_dict[vert].x) + " " + str(vert_dict[vert].y) + " " + str(vert_dict[vert].z) + " ") 
    file.write(str(color[0]) + " " + str(color[1]) + " " + str(color[2]) + "\n")
        
def write_edge_points(strong, weak, vert_dict, face_dict, target_file):
    with open(target_file +"_EdgePoints.ply", "w") as f:
        nb_points = len(vert_dict)
        nb_faces = len(face_dict)
        write_header(f, nb_points, nb_faces)

        for vert in vert_dict:
            if vert in strong:
                write_vertex(f, vert, vert_dict, color=[255,0,0])
            elif vert in weak:
                write_vertex(f, vert, vert_dict, color=[0,0,255])
            else:
                write_vertex(f, vert, vert_dict, color=[187,187,187])
        for face in face_dict.keys():
            ind = list(face_dict[face].indices)
            f.write("3 " + str(ind[0]) + " " + str(ind[1]) + " " + str(ind[2]) + "\n")

In [23]:
def edges(artifact, high, low, index = 7):
    data_conf = Morse()
    data_conf.load_mesh_ply("Data/"+artifact+".ply", quality_index=index, inverted=True)

    outfilename =  "Data/results/" + artifact
    #### CONFORMING
    data_conf.load_labels("Data/"+artifact+"_labels.txt")
    data_conf.ConformingGradient()
    data_conf.ExtractMorseComplex()
    data_conf.ReduceMorseComplex(10, conforming=True)

    strong, weak = get_salient_edge_indices(data_conf.maximalReducedComplex, high, low, 
                                                      data_conf.Vertices, data_conf.Edges, data_conf.Faces)

    write_edge_points(strong, weak, data_conf.Vertices, data_conf.Faces, outfilename)
                                                

In [24]:
edges('700', 0.06, 0.02)
edges('801', 0.06, 0.04, index=3)
edges('4501', 0.1, 0.07)

load_mesh_ply took 22.45025 seconds to execute!
Time read labels in txt file: 0.12104100012220442
Time ProcessLowerStar: 6.4175768999848515
ExtractMorseComplex took 4.62482 seconds to execute!
Persistence was high enough that this complex is maximally reduced.
ReduceMorseComplex took 7.47769 seconds to execute!
load_mesh_ply took 20.66929 seconds to execute!
Time read labels in txt file: 0.09974079998210073
Time ProcessLowerStar: 6.025972499977797
ExtractMorseComplex took 3.85070 seconds to execute!
Persistence was high enough that this complex is maximally reduced.
ReduceMorseComplex took 6.37018 seconds to execute!
load_mesh_ply took 17.32096 seconds to execute!
Time read labels in txt file: 0.11164380004629493
Time ProcessLowerStar: 7.0062539000064135
ExtractMorseComplex took 6.97315 seconds to execute!
Persistence was high enough that this complex is maximally reduced.
ReduceMorseComplex took 6.07637 seconds to execute!
